### PR TITLE
Feature/update lintball

### PR DESCRIPTION
**Pull request by ghost**
- separated all linters
- added json lint
- added lintignore 
- added test files to lint against to ensure some level of quality
- added versioning
- added cloudformation to create ECR repo 
- published to ECR
- updated readme
- instructions on pre-commit hooks for auto linting

### CODE REVIEW

**Review by 17twenty — COMMENTED** [review]
I'd recommend filling out or removing TODOs - I'd up the docs (naked URLs without context aren;t helpful) and then I'd add a lot of usage examples. 
I'd consider whether the `lib` directory makes sense to overload with all the things (`/usr/lib` would've been better if it's that you were adhering to Linux FSB but even then, i'd argue configs should go in `/etc`) 

Excellent work tho lads!

File: README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  + e.g. As a local pre-commit hook or a build pipeline (aws code build/jenkins/etc...), where a list of changed filenames is passed into the Lintball process)
+
++ Option 2 - Pass the details of the git repo / git branch / git commit / etc... to the Lintball container. Lintball will clone the repo locally, checkout the branch etc... and apply the `Linters`.
+
+(nb: Option 2 is provided as a temporary fix to get around this [issue](https://github.com/aws/aws-codebuild-docker-images/issues/76) )
+Details of using option 2 are provided at the end of this README.
+
+## Option 1
+
+### Getting Started
```
Comment (balloonpopper) [inline README.md]: Need to call out here how to get started for local usage as opposed to pipeline usage

File: README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
```
Comment (balloonpopper) [inline README.md:36]: Need pre-req's installed
* docker
* docker-compose

File: lib/common.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash -e
+
+export SEP="====================================================================="
+
+log(){
+  echo "$(date +%Y-%m-%dT%H:%M:%S%z) INFO: $(basename "${0}") [$$] ${*}"
```
Comment (balloonpopper) [inline lib/common.sh:6]: Spacing is inconsistent in shell scripts - it's 2 spaces in some places, 4 spaces in others

**Review by balloonpopper — COMMENTED** [review]
Can you make it echo out the number of files it processed? 
Ideally it would be good to have 3 different print options
1) How many files were processed
2) How many files of each type were  processed
3) Every file with a pass / fail indicator

**Comment by balloonpopper** [comment]
Throwing lots of files at this causes breakages
eg. Ïn a big directory : 
docker run -v "$PWD:/scan" -e DEBUG="false" --rm lintball:1.0.0 "$(find . -print0 | perl -e 'local $/="\0";print join(" ",map{s/(?=["'"'"' ])/\\/gr}<STDIN>);')"

This returns : bash: warning: command substitution: ignored null byte in input
bash: /bin/docker: Argument list too long

File: lib/githooks/pre-commit [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+determine_git_revision()
+{
+  # Work out what to diff git_against, really HEAD will work for any established repository.
+  if git rev-parse --verify HEAD >/dev/null 2>&1
+  then
+    git_against=HEAD
+  else
+    # Initial commit: diff git_against an empty tree object
+    git_against=4b825dc642cb6eb9a060e54bf8d69288fbee4904
```
Comment (balloonpopper) [inline lib/githooks/pre-commit]: What's the significance of this specific git reference?

File: lib/githooks/pre-commit [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+
+########
+# MAIN #
+###############################################################################
+check_prerequisites
+determine_git_revision
+
+RC=0
```
Comment (balloonpopper) [inline lib/githooks/pre-commit:64]: Duplicate RC=0

File: lib/githooks/pre-commit [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+declare CHANGED_FILES=""
+CHANGED_FILES="$(git diff-index --cached --name-only "$git_against")"
+
+dump_changed_files ${git_against}
+
+#TODO - @muz - source in the lintball version from an external file (shared with the Makefile, to ensure version is in sync)
+docker run -v "$PWD:/scan" -e DEBUG="true" --rm lintball:1.0.0 "${CHANGED_FILES}"
+
+echo "RC=${RC}"
```
Comment (balloonpopper) [inline lib/githooks/pre-commit]: RC isn't set
Also, if docker returns non-zero, this isn't going to fire due to bash -e

File: README.md (hunk 2 above)
Comment (ghost) [reply README.md:36]: added pre-req

File: lib/common.sh (hunk 3 above)
Comment (ghost) [reply lib/common.sh:6]: fixed file spacing

File: test/test_files/example.cfn [hunk 7]
```diff
@@ -0,0 +1,516 @@
+---
+AWSTemplateFormatVersion: "2010-09-09"
+Description:
+  CodePipeline and CodeBuild projects to deploy CIP applications
```
Comment (micklove) [inline test/test_files/example.cfn]: Should remove this file, or "sanitise" the names.

**Review by micklove — APPROVED** [review]

**Pull request merged**